Project 9(api)

cjrodriguez12 opened the pull request.

cjrodriguez12 committed:
merge

cjrodriguez12 committed:
API set up

loading profile info and  working on setting profile to server

cjrodriguez12 committed:
stage 3 editing profile

cjrodriguez12 committed:
updating profile error

sucessfully updates name and description but still recieving error from index line 95

cjrodriguez12 committed:
Working On Post Method

POst method for cards

cjrodriguez12 committed:
Post cards

Doesn't save after refreshing page

cjrodriguez12 committed:
rendering cards through api

cjrodriguez12 committed:
gallery card delete attempt

cjrodriguez12 committed:
Card Delete Success

tagged Card element with Id and isLiked from Api created function to delete card by id then remove from UI

cjrodriguez12 committed:
IsLikes function attempt

isLiked chnages from true/false in api but not reflected in UI

cjrodriguez12 committed:
isLiked now adds/remove upon card render

cjrodriguez12 committed:
UX cleanup attempt

submit button changes to "saving...'
on submit
changes back after server reQuest is complete utilizing then/catch/finally

cjrodriguez12 committed:
toggle button state/form reset fix

gennady-bars reviewed (commented):
Dear Charlie , (**Please, expand the general comment ↓**)

Unfortunately I have to reject the project:


* When I like new cards there are errros https://disk.yandex.ru/i/gEI8Hm4rtAgeVQ
* When I delete new cards there are errors https://disk.yandex.ru/i/vaF0gMql4rwnPw

cjrodriguez12 committed:
new cards didn't pass card I.D.

fixed by returning handled response in postcards api

gennady-bars commented on src/components/Api.js:
```diff
@@ -0,0 +1,105 @@
+  getInitialCards() {
+    return fetch(`${this._baseUrl}/cards`, {
+      headers: this._headers,
+    }).then(this._handleResponse);
```
**Could be improved**

You can make a special method for fetching and checking responses **not to duplicate** it in every request:

```js

  _request(url, options) {
    return fetch(url, options).then(this._handleResponse)
  }

```

And now you can safely replace `fetch` with `this._request` and remove repeated checking. And you don’t need to change anything else.

gennady-bars commented on src/components/Api.js:
```diff
@@ -0,0 +1,105 @@
+export class Api {
+  constructor() {
+    // constructor body
+    this._baseUrl = "https://around-api.en.tripleten-services.com/v1";
+    this._headers = {
+      authorization: "2e65c592-5cc5-4cb0-a6bf-23fa612e6f57",
+      "Content-Type": "application/json",
+    };
```
Please, don't hardcode this data inside `Api`

According to the task you need to pass full `headers` and `baseUrl` into the **constructor** of `Api`. See here https://disk.yandex.ru/i/J93dbJp7eypXKQ .

gennady-bars commented on src/components/Card.js:
```diff
@@ -26,12 +39,19 @@ export default class Card {
   /**Event Handlers */
-  _handleLike() {
+  handleLike() {
     this._cardElement
       .querySelector(".gallery__card-button")
```
You've found the like button in line 69 and it's accessible here as `this._likeButton`


```js
  handleLike() {
    this._likeButton.classList.toggle("gallery__card-button_active");
  }

```

gennady-bars commented on src/components/Card.js:
```diff
@@ -1,23 +1,36 @@
   }
   _setEventListeners() {
     //like btn
     this._cardElement
       .querySelector(".gallery__card-button")
```
You've found the like button in line 69 and it's accessible here as `this._likeButton`

`Class fields` (the so called **class variables** with `this`) are accessible inside any method of the class. You shouldn’t pass **class fields** into the call of any method or search them again because you can use class fields inside all methods.

gennady-bars commented on src/components/Card.js:
```diff
@@ -46,11 +66,13 @@ export default class Card {
     // get card view
     this._card = this.getTemplate();
+    this._likeButton = this._card.querySelector(".gallery__card-button");
     this._card.querySelector(".gallery__card-image").src = this._link;
```
`this._card.querySelector(".gallery__card-image")`

**Could be improved**

All elements that are found every time you call a method should be defined as **class fields** in the constructor, `_setEventListeners` (or in `getView`) only **1 time** not to waste resources on searching them again and again.

`Class fields` (the so called **class variables** with `this`) are accessible inside any method of the class. You shouldn’t pass **class fields** into the call of any method or search them again because you can use class fields inside all methods.
There can be `this._inputs`, `this._submitButton`, `this._cardImage`, `this._form` , `this._likeButton` and so on.

```js
  getView() {
    this._card = this.getTemplate();
    this._likeButton = this._card.querySelector(".gallery__card-button");
    this._cardImage = this._card.querySelector(".gallery__card-image")
    this._trashButton = this._card.querySelector(".gallery__card-delete")

```

gennady-bars commented on src/components/PopupWithForm.js:
```diff
@@ -5,18 +5,23 @@ export class PopupWithForm extends Popup {
       const modalInputs = this._getInputValues();
-      this._handleModalSubmit(modalInputs);
+      e.preventDefault();
+      this._popUpForm.querySelector(".modal__form-button").innerText =
```
`this._popUpForm.querySelector(".modal__form-button")`

**Could be improved**

All elements that are found every time you call a method should be defined as **class fields** in the constructor, `_setEventListeners` (or in `generateCard`) only **1 time** not to waste resources on searching them again and again.

`Class fields` (the so called **class variables** with `this`) are accessible inside any method of the class. You shouldn’t pass **class fields** into the call of any method or search them again because you can use class fields inside all methods.
There can be `this._inputs`, `this._submitButton`, `this._cardImage`, `this._form` , `this._likeButton` and so on.

gennady-bars replied:
**Could be improved**

It’s better to make method `renderLoading` inside class `PopupWithForm`. You should find the button once in the constructor and change the text:  

```js
  constructor({ popupSelector, handleSubmit }) {
    super(popupSelector);
    // fix the initial button text only once in the constructor
    this._submitBtnText = this._submitBtn.textContent
  }
  // add 2 params: isLoading and loadingText with a default text
  renderLoading(isLoading, loadingText='Saving...') {
    if (isLoading) {
      this._submitBtn.textContent = loadingText;
    } else {
	// here we return back the initial text. So, you don’t need to bother yourself about it
      this._submitBtn.textContent = this._submitBtnText;
    }
  }

```

gennady-bars commented on src/components/PopupWithForm.js:
```diff
@@ -5,18 +5,23 @@ export class PopupWithForm extends Popup {
     });
   }
+  setSubmitAction(callBackfn) {
+    this._handleModalSubmit = callBackfn;
```
Nice way of updating the function

gennady-bars commented on src/components/PopupWithForm.js:
```diff
@@ -5,18 +5,23 @@ export class PopupWithForm extends Popup {
   }
+  setSubmitAction(callBackfn) {
+    this._handleModalSubmit = callBackfn;
+  }
```
**Could be improved**

You can make ` setInputValues` in class `PopupWithForm`, which can insert data into inputs:

```js

  setInputValues(data) {
    this._inputList.forEach((input) => {
      // here you insert the `value` by the `name` of the input
      input.value = data[input.name];
    });
  }



```

So, this way you’ll not have to search the inputs of the profile in `index.js`

gennady-bars commented on src/pages/index.js:
```diff
@@ -21,91 +34,155 @@ popUpImageModal.setEventListeners();
 }
-
+//Popup for deleting cards from API along with handler
+const dltPopup = new PopupWithForm(constants.selectors.deleteSelector);
```
`dltPopup`

Please, use word `delete` or `confirm` here to make it clear what popup it is


`confirmPopup` or `deletePopup` or `confirmationPopup`

gennady-bars commented on src/pages/index.js:
```diff
@@ -21,91 +34,155 @@ popUpImageModal.setEventListeners();
+      .then(() => {
+        dltPopup.closeModal();
+        card.deleteUI();
+      })
```
You need to `catch` possible errors at the very end of any server request.

**Please, correct it everywhere**

gennady-bars commented on src/pages/index.js:
```diff
@@ -21,91 +34,155 @@ popUpImageModal.setEventListeners();
+//Handler for the Like button
+function handleToggleLikes(id, isLiked, card) {
+  if (isLiked === false) {
+    card.handleLike();
+    isLiked = true;
```
```js

    card.handleLike();
    isLiked = true;
```
You need to **toggle likes** only in block `then` after a successful response from the server because if there is an error the user will not know about that and will see the toggled heart color.

gennady-bars commented on src/pages/index.js:
```diff
@@ -21,91 +34,155 @@ popUpImageModal.setEventListeners();
+    isLiked = true;
+    api.toLike(id);
+  } else {
+    card.handleLike();
+    isLiked = false;
```
You need to **toggle likes** only in block `then` after a successful response from the server because if there is an error the user will not know about that and will see the toggled heart color.

gennady-bars commented on src/pages/index.js:
```diff
@@ -21,91 +34,155 @@ popUpImageModal.setEventListeners();
-<<<<<<< Updated upstream
 editFormValidator.enableValidation();
+//avatar Form - link
+const avatarFormValidator = new FormValidator(
```
**Could be improved**

You can **universally** create instances of validators for all forms in the project storing them inside one object `formValidators`.  And then you can take any validator using attribute `name` of the form where you need to disable/enable the submit button or remove errors.

```js
const formValidators = {}

// enable validation
const enableValidation = (config) => {
  const formList = Array.from(document.querySelectorAll(config.formSelector))
  formList.forEach((formElement) => {
    const validator = new FormValidator(config, formElement)
    // here you get the name of the form
    const formName = formElement.getAttribute('name')

   // here you store a validator by the `name` of the form
    formValidators[formName] = validator;
   validator.enableValidation();
  });
};

enableValidation(config);

```

And now you can use them for disabling buttons or clearing errors:

```js

formValidators[ profileForm.getAttribute('name') ].resetValidation()

// or you can use a string – the name of the form (you know it from `index.html`)

formValidators['profile-form'].resetValidation()

```

gennady-bars commented on src/pages/index.js:
```diff
@@ -21,91 +34,155 @@ popUpImageModal.setEventListeners();
-avatarFormValidator.enableValidation();
->>>>>>> Stashed changes
 // call queryselector on input inside form /set values
 function handleFormFill(userInfoList) {
```
There are 3 popups in the project with buttons and 2 forms. You need to **name** the elements the way all could understand what they are. And the functions as well


`fillProfileInputs`

gennady-bars commented on src/pages/index.js:
```diff
@@ -21,91 +34,155 @@ popUpImageModal.setEventListeners();
+      editFormValidator.toggleButtonState();
+    })
+    .finally(() => {
+      popUpForm.querySelector(".modal__form-button").innerText = "Save";
```
**Could be improved**

If it’s interesting for you here is how we can make a universal function for handling any submit. We can get rid of such duplicating as **loading effect**, **closing** and **catching errors**


```js

// You can make a universal function that accepts a request, popup instance and optional loading text  
function handleSubmit(request, popupInstance, loadingText = "Saving...") {

  // here we change the button text
  popupInstance.renderLoading(true, loadingText);
  request()
    .then(() => {
      // We need to close only in `then`
      popupInstance.close()
    })
      // we need to catch possible errors
      // console.error is used to handle errors if you don’t have any other ways for that
     .catch(console.error)
    // in `finally` we need to return the initial button text back in any case
    .finally(() => {
      popupInstance.renderLoading(false);
    });
}



```


You need to pass the **request** into the call of `handleSubmit` like that:



```js

// here is an example of the profile form handling
function handleProfileFormSubmit(inputValues) {
  // we create a function that returns a promise
  function makeRequest() {
    // `return` lets us use a promise chain `then, catch, finally` inside `handleSubmit`
    return api.editProfile(inputValues).then((userData) => {
      userInfo.setUserInfo(userData)
    });
  }
  // Here we call the function passing the request, popup instance and if we need some other loading text we can pass it as the 3rd argument
  handleSubmit(makeRequest, profilePopup);
}



```

gennady-bars commented on src/pages/index.js:
```diff
@@ -21,91 +34,155 @@ popUpImageModal.setEventListeners();
+    })
+    .finally(() => {
+      popUpForm.querySelector(".modal__form-button").innerText = "Save";
+      profilePopup.closeModal();
```
`      profilePopup.closeModal();
`

You need to close popups only in blocks `then` after a **successful response** from the server so that when there is an error it would be opened (the inputs will not be cleared and the user could send the data again to the server) and if you change the button text you could see the change as well

**Please, correct it everywhere**

gennady-bars commented on src/pages/index.js:
```diff
@@ -21,91 +34,155 @@ popUpImageModal.setEventListeners();
+    .updateAvatar(modalInputs)
+    .then(() => {
+      const link = modalInputs.Url;
+      constants.profileAvatar.src = link;
```
`      constants.profileAvatar.src = link;
`

Please, delete it

`newUserInfo.setAvatar(link);` should set the avatar

gennady-bars replied:
Class `UserInfo` should set the entire user’s data (including the **avatar**)

gennady-bars commented on src/pages/index.js:
```diff
@@ -21,91 +34,155 @@ popUpImageModal.setEventListeners();
+api.loadInfo().then((data) => {
+  const { name, about, avatar } = data;
+  newUserInfo.setUserInfo(name, about);
+  constants.profileAvatar.src = avatar;
```
`      constants.profileAvatar.src = avatar;
`

Please, delete it

`newUserInfo.setAvatar(avatar);` should set the avatar

gennady-bars commented on src/pages/index.js:
```diff
@@ -21,91 +34,155 @@ popUpImageModal.setEventListeners();
+      avatarFormValidator.toggleButtonState();
+    })
+    .finally(() => {
+      popUpForm.querySelector(".modal__form-button").innerText = "Save";
```
Good job that you return the button text in `finally`

gennady-bars reviewed (commented):
Dear Charlie , (**Please, expand the general comment ↓**)

Good job but there are some remarks:


* You need to `catch` possible errors at the very end of any server request.
* You need to **toggle likes** only in block `then` after a successful response from the server because if there is an error the user will not know about that and will see the toggled heart color.
* There are 3 popups in the project with buttons and 2 forms. You need to **name** the elements the way all could understand what they are. And the functions as well
* You need to close popups only in blocks `then` after a **successful response** from the server so that when there is an error it would be opened (the inputs will not be cleared and the user could send the data again to the server) and if you change the button text you could see the change as well
* Class `UserInfo` should set the entire user’s data (including the **avatar**)



Please, correct everything and your project will be accepted.

Good luck with the refactoring

cjrodriguez12 committed:
reviewer symantics and corrections

completed the following
You need to catch possible errors at the very end of any server request.
You need to toggle likes only in block then after a successful response from the server because if there is an error the user will not know about that and will see the toggled heart color.
There are 3 popups in the project with buttons and 2 forms. You need to name the elements the way all could understand what they are. And the functions as well
You need to close popups only in blocks then after a successful response from the server so that when there is an error it would be opened (the inputs will not be cleared and the user could send the data again to the server) and if you change the button text you could see the change as well
Class UserInfo should set the entire user’s data (including the avatar)

cjrodriguez12 committed:
reviewer syamntics anD some improvements

gennady-bars reviewed (commented):
Everything is almost fine but:

* When I edit the profile data the avatar disappears https://disk.yandex.ru/i/-PUvuy4FyTNszQ 
* When I edit the avatar nothing is disapleayed  https://disk.yandex.ru/i/g27Dg8FFdPuzqw

cjrodriguez12 committed:
editing avatar  error

fixed with setAvatar in class UserInfo

cjrodriguez12 commented:
Thanks for  your help! your review was insightful and I wanted to implement more of the refactoring that you had sent but the project is way past due and the holidays are here so don't have as much time currently. In the future i hope i can get to it

gennady-bars reviewed (commented):
One more step and everything will be fine:


* When I edit the profile data the avatar disappears and there is an error https://disk.yandex.ru/i/WzJZrdcaeCn8ew 
* Class `UserInfo` should set the entire user’s data (including the **avatar**)

cjrodriguez12 committed:
symantics and profile error

When I edit the profile data the avatar disappears and there is an error https://disk.yandex.ru/i/WzJZrdcaeCn8ew
Class UserInfo should set the entire user’s data (including the avatar)

cjrodriguez12 commented:
sorry I thought i fixed what you asked of me but i clearly misunderstood. I believe it's fixed now.

gennady-bars reviewed (approved):
You’ve done a great job.

Your project has been accepted.

Good luck
